src/Xamarin.Android.Build.Tasks/Tasks/LinkApplicationSharedLibraries.cs

```diff
@@ -0,0 +1,260 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using ThreadingTasks = System.Threading.Tasks;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+using Xamarin.Android.Tools;
+
+namespace Xamarin.Android.Tasks
+{
+	public class LinkApplicationSharedLibraries : AsyncTask
+	{
+		sealed class Config
+		{
+			public string LinkerPath;
+			public string LinkerOptions;
+			public string OutputSharedLibrary;
+		}
+
+		sealed class InputFiles
+		{
+			public List<string> ObjectFiles;
+			public string OutputSharedLibrary;
+		}
+
+		[Required]
+		public ITaskItem[] ObjectFiles { get; set; }
+
+		[Required]
+		public ITaskItem[] ApplicationSharedLibraries { get; set; }
+
+		[Required]
+		public bool DebugBuild { get; set; }
+
+		[Required]
+		public string AndroidNdkDirectory { get; set; }
+
+		public override bool Execute ()
+		{
+			try {
+				return DoExecute ();
+			} catch (Exception e) {
+				Log.LogCodedError ("XA3001", $"{e}");
+				return false;
+			}
+		}
+
+		bool DoExecute ()
+		{
+			Yield ();
+			try {
+				var task = ThreadingTasks.Task.Run ( () => RunParallelLinker (), Token);
+				task.ContinueWith (Complete);
+
+				base.Execute ();
+
+				if (!task.Result)
+					return false;
+			} finally {
+				Reacquire ();
+			}
+
+			return !Log.HasLoggedErrors;
+		}
+
+		bool RunParallelLinker ()
+		{
+			try {
+				ThreadingTasks.ParallelOptions options = new ThreadingTasks.ParallelOptions {
+					CancellationToken = Token,
+					TaskScheduler = ThreadingTasks.TaskScheduler.Default,
+				};
+
+				ThreadingTasks.Parallel.ForEach (GetLinkerConfigs (), options,
+								 config => {
+									 if (!RunLinker (config)) {
+										 LogCodedError ("XA3001", $"Could not link native shared library: {Path.GetFileName (config.OutputSharedLibrary)}");
+										 Cancel ();
+										 return;
+									 }
+								 }
+				);
+			} catch (OperationCanceledException) {
+				return false;
+			}
+
+			return true;
+		}
+
+		bool RunLinker (Config config)
+		{
+			var stdout_completed = new ManualResetEvent (false);
+			var stderr_completed = new ManualResetEvent (false);
+			var psi = new ProcessStartInfo () {
+				FileName = config.LinkerPath,
+				Arguments = config.LinkerOptions,
+				UseShellExecute = false,
+				RedirectStandardOutput = true,
+				RedirectStandardError = true,
+				CreateNoWindow = true,
+				WindowStyle = ProcessWindowStyle.Hidden,
+			};
+
+			string targetDir = Path.GetDirectoryName (config.OutputSharedLibrary);
+			if (!Directory.Exists (targetDir))
+				Directory.CreateDirectory (targetDir);
+
+			string linkerName = Path.GetFileName (config.LinkerPath);
+			LogDebugMessage ($"[Native Linker] {psi.FileName} {psi.Arguments}");
+			using (var proc = new Process ()) {
+				proc.OutputDataReceived += (s, e) => {
+					if (e.Data != null)
+						OnOutputData (linkerName, s, e);
+					else
+						stdout_completed.Set ();
```

Comment:
Considering that this code pattern is present in *at least* two different tasks *within this PR*, could we please refactor out the logic to "something" a'la:

```csharp
partial class MonoAndroidHelper {
    public static bool ExecuteProcess (ProcessStartInfo psi, Action<string, object, DataReceivedEventArgs> onOutputData, Action<string, object, DataReceivedEventArgs> onErrorData);
}
```

As you say, `Process` is an API which is *easy* to get wrong.  We should thus share as much code as possible, to reduce the likelihood of screwing it u p.

Reply:
I agree we should refactor it, but not in this PR. It should be a separate effort as we use `S.D.Process` in quite a few places and this PR would unnecessarily grow because of all the modifications, and with changes that are not related to the subject of this PR.

Reply:
My thinking was that since this `S.D.Process` code is *already duplicated* within this PR, adding a helper method would *reduce* the size of this PR, even if it was *only* used for these newly added locations.

Reply:
@grendello if you could make a "blessed" `System.Diagnostics.Process` helpers and just use in the tasks in this PR. We could start porting the other tasks, which would help a lot.